### PR TITLE
refactor(app-shell): redo project layout, build system, test runner, package layout, and exports

### PR DESCRIPTION
**Pull request by alxhub**
Also moves shell-parser to an experimental directory until it can be updated.

### CODE REVIEW

File: app-shell/gulpfile.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+  // Set up ts-node to enable loading of TypeScript files.
+require('ts-node').register({
+  noProject: true,
+});
+
+// Trampoline into gulpfile.ts.
+require('./gulpfile.ts');
```
Comment (jeffbcross) [inline app-shell/gulpfile.js]: I recently learned that with `ts-node`, you don't need this file anymore. You'll have to delete `declare var require` from gulpfile.ts, but then you can just run `gulp build`.

File: app-shell/package.json [hunk 2]
```diff
@@ -1,51 +1,47 @@
...
-    "karma": "^0.13.15",
-    "karma-chrome-launcher": "^0.2.3",
-    "karma-jasmine": "^0.3.8",
-    "parse5": "2.1.5",
-    "protractor": "^3.3.0",
-    "ts-node": "^0.5.5",
-    "tslint": "^3.6.0",
-    "typescript": "^2.0.2",
-    "typings": "^1.0.4"
+    "@angular/common": "^2.0.0",
```
Comment (jeffbcross) [inline app-shell/package.json]: The Angular, Zone and RxJS dependencies should be peer dependencies in the published package.json. Fine to move them to dev dependencies here though.

File: app-shell/src/app/index.ts [hunk 3]
```diff
@@ -1,28 +1,3 @@
...
-    return {
-      ngModule: AppShellModule,
-      providers: [APP_SHELL_RUNTIME_PROVIDERS, APP_SHELL_BUILD_PROVIDERS]
-    };
-  }
-}
-
+export * from './module';
+export * from './prerender';
+export * from './shell';
```
Comment (jeffbcross) [inline app-shell/src/app/index.ts]: Newline plz

File: app-shell/src/app/module.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    }
+  }
+
+  static runtime(): ModuleWithProviders {
+    return {
+      ngModule: AppShellModule,
+      providers: [{provide: IS_PRERENDER, useValue: false}],
+    };
+  }
+}
```
Comment (jeffbcross) [inline app-shell/src/app/module.ts]: Newline plz

File: app-shell/src/app/prerender.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+import {OpaqueToken} from '@angular/core';
+
+export const IS_PRERENDER = new OpaqueToken('IsPrerender');
```
Comment (jeffbcross) [inline app-shell/src/app/prerender.ts]: Newline plz

File: app-shell/src/index.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+export * from './app';
```
Comment (jeffbcross) [inline app-shell/src/index.ts]: Newline plz

File: app-shell/tsconfig.es5.json [hunk 7]
```diff
@@ -1,24 +1,20 @@
...
+	},
   "files": [
-    "main.ts",
-    "main-app-shell.ts",
-    "typings.d.ts"
+    "src/index.ts",
+    "src/unit_tests.ts"
   ]
-}
+}
```
Comment (jeffbcross) [inline app-shell/tsconfig.es5.json]: Newline plz

File: app-shell/tsconfig.esm.json [hunk 8]
```diff
@@ -1,22 +1,24 @@
...
+    "experimentalDecorators": true,
+    "declaration": true,
+    "outDir": "tmp/esm",
     "sourceMap": true,
-    "inlineSources": true
+    "moduleResolution": "node",
+    "rootDir": ".",
+    "baseUrl": ".",
+    "paths": {
+      "@angular/service-worker/worker": ["src/worker"]
```
Comment (jeffbcross) [inline app-shell/tsconfig.esm.json]: Why is this necessary?

File: app-shell/tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+    "outDir": "dist",
+    "sourceMap": false,
+    "moduleResolution": "node",
+    "rootDir": ".",
+    "baseUrl": "."
+	},
+  "exclude": [
+    "node_modules"
+  ]
+}
```
Comment (jeffbcross) [inline app-shell/tsconfig.json]: Newline plz

File: app-shell/package.json (hunk 2 above)
Comment (alxhub) [inline app-shell/package.json]: I think `@angular/core` is the only real peer dep.

File: app-shell/tsconfig.esm.json (hunk 8 above)
Comment (alxhub) [inline app-shell/tsconfig.esm.json]: Copy-paste error - gone.

File: app-shell/tsconfig.json (hunk 9 above)
Comment (alxhub) [inline app-shell/tsconfig.json]: Done.

File: app-shell/tsconfig.es5.json (hunk 7 above)
Comment (alxhub) [inline app-shell/tsconfig.es5.json]: Done.

File: app-shell/src/index.ts (hunk 6 above)
Comment (alxhub) [inline app-shell/src/index.ts]: Done.

File: app-shell/src/app/prerender.ts (hunk 5 above)
Comment (alxhub) [inline app-shell/src/app/prerender.ts]: Done.

File: app-shell/src/app/module.ts (hunk 4 above)
Comment (alxhub) [inline app-shell/src/app/module.ts]: Done.

File: app-shell/src/app/index.ts (hunk 3 above)
Comment (alxhub) [inline app-shell/src/app/index.ts]: Done.
